File: findMinimumInRotatedSortedArrayII.py
class Solution:
    # @param num, a list of integer
    # @return an integer
    def findMin(self, num):
        return self.findMin_interval(num, 0, len(num)-1)

    def findMin_interval(self, num, start, end):
        if start == end:
            return num[start]
        if num[start] < num[end]:
            return num[start]
        mid = start + (end - start) / 2
        if num[start] == num[end]:
            return self.findMin_interval(num, start+1, end)
        if num[mid] >= num[start]:
            return self.findMin_interval(num, mid+1, end)
        return self.findMin_interval(num, start, mid)

File: findMinimumInRotatedSortedArrayII.java
public class Solution {
    public int findMin(int[] num) {
        return findMin(num, 0, num.length-1);
    }
    private int findMin(int[] num, int start, int end) {
        if (start == end)
            return num[start];
        if (num[start] < num[end])
            return num[start];
        int mid = start + (end - start) / 2;
        if (num[start] == num[end])
            return findMin(num, start+1, end);
        if (num[mid] >= num[start])
            return findMin(num, mid+1, end);
        return findMin(num, start, mid);
    }
}
